File: tools/new_leaf.py
#!/usr/bin/env python3
"""Stamp a NEW conforming fak leaf — the golden path for modular growth.

fak grows by adding leaves: a new package under `fak/internal/` that depends only
on lower layers and (if it sits on the request path) registers itself against the
frozen ABI from `init()`. The architecture makes that the ONLY extension mechanism
(see fak/ARCHITECTURE.md), and architest mechanically enforces it. This script makes
the conforming shape the path of LEAST resistance: one command produces a leaf that
is green by construction — declared in the tier table, optionally slotted into the
defconfig, and carrying a passing test — so growth can never silently erode the
layered-DAG contract.

It performs three edits, each idempotent and refused if already done:
  1. creates fak/internal/<name>/{doc.go,<name>.go,<name>_test.go}
  2. inserts the package's tier into internal/architest (above the new-leaf marker)
  3. with --register, adds the blank-import line to internal/registrations (defconfig)

Pure stdlib; off the request path (DIRECTION.md tooling seam).

Usage:
  python tools/new_leaf.py canon2 --tier foundation
  python tools/new_leaf.py fedtrust --tier composer --register --summary "federated trust gate"
  python tools/new_leaf.py foo --tier mechanism --dry-run
"""
from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path

NAME_RE = re.compile(r"^[a-z][a-z0-9]*$")
TIERS = {"root": 0, "foundation": 1, "mechanism": 2, "composer": 3, "integrator": 4}
MOD = "github.com/anthony-chaudhary/fak/internal"
TIER_MARKER = "// new-leaf:tier"


def repo_root() -> Path:
    return Path(__file__).resolve().parent.parent


def doc_go(name: str, tier: str, n: int, summary: str) -> str:
    return (
        f"// Package {name} is {summary}.\n"
        f"//\n"
        f"// Tier: {tier} ({n}) — see internal/architest. This package may import only\n"
        f"// packages whose tier is <= {n}; an upward import fails the architest gate.\n"
        f"// See fak/GROWTH.md for the layering contract and how a leaf bakes in.\n"
        f"package {name}\n"
    )


def impl_go(name: str, register: bool) -> str:
    head = f"package {name}\n"
    if register:
        head += f'\nimport "{MOD}/abi"\n'
    body = (
        "\n// Ready reports that the leaf is wired. Replace this placeholder with the\n"
        "// real surface (the type/constructor this package exists to provide).\n"
        "func Ready() bool { return true }\n"
    )
    if register:
        body += (
            "\n// init registers this leaf's driver against the frozen ABI before the\n"
            "// kernel boots. Pick the Register* that matches the leaf's role — see the\n"
            "// table in fak/ARCHITECTURE.md (RegisterAdjudicator / RegisterFastPath /\n"
            "// RegisterOp / RegisterReason / RegisterEngine / ...).\n"
            "func init() {\n"
            "\t_ = abi.ABIMinor // TODO: replace with the real abi.Register*(...) for this leaf\n"
            "}\n"
        )
    return head + body


def test_go(name: str) -> str:
    return (
        f"package {name}\n\n"
        'import "testing"\n\n'
        "func TestReady(t *testing.T) {\n"
        "\tif !Ready() {\n"
        '\t\tt.Fatal("Ready() should report true for the generated skeleton")\n'
        "\t}\n"
        "}\n"
    )


def insert_before_marker(text: str, marker: str, line: str) -> str:
    """Insert `line` (with trailing newline) immediately before the line containing
    `marker`, preserving its indentation."""
    out = []
    done = False
    for ln in text.splitlines(keepends=True):
        if not done and marker in ln:
            out.append(line)
            done = True
        out.append(ln)
    if not done:
        raise ValueError(f"marker {marker!r} not found")
    return "".join(out)


def add_registration(text: str, name: str) -> str:
    """Insert the blank-import line before the final closing paren of the import block."""
    imp = f'\t_ "{MOD}/{name}"\n'
    if imp in text:
        return text  # idempotent
    idx = text.rfind("\n)")
    if idx == -1:
        raise ValueError("could not find import block close ')' in registrations")
    return text[: idx + 1] + imp + text[idx + 1 :]


def insert_before_all_markers(text: str, marker: str, line: str) -> str:
    """Insert `line` before EVERY line containing `marker` (the lane marker appears in
    both the concurrent and autopick arrays)."""
    out, hits = [], 0
    for ln in text.splitlines(keepends=True):
        if marker in ln:
            out.append(line)
            hits += 1
        out.append(ln)
    if hits == 0:
        raise ValueError(f"marker {marker!r} not found")
    return "".join(out)


def add_leaf_lane(text: str, name: str) -> str:
    """Add the leaf's per-leaf concurrency lane to dos.toml: its name to the concurrent
    and autopick arrays (the `# new-leaf:lane` markers) and its disjoint tree to
    [lanes.trees] (the `# new-leaf:tree` marker). Idempotent."""
    # The lane tree is matched against the LIVE workspace, whose Go module is the
    # REPOSITORY ROOT (AGENTS.md) -- the real path is internal/<name>/, NOT
    # fak/internal/<name>/. Emitting the `fak/` form makes the glob match zero
    # files, so the arbiter cannot detect a collision on the leaf (dos-effective-
    # usage audit, 2026-06-22). Idempotency keys on the real-layout tree line (and
    # still treats a legacy `fak/internal/<name>/**` entry as already-present).
    if f'["internal/{name}/**"]' in text or f'fak/internal/{name}/**' in text:
        return text
    text = insert_before_all_markers(text, "# new-leaf:lane", f'  "{name}",\n')
    text = insert_before_marker(text, "# new-leaf:tree", f'{name} = ["internal/{name}/**"]\n')
    return text


def main() -> int:
    ap = argparse.ArgumentParser(description="Stamp a new conforming fak leaf.")
    ap.add_argument("name", help="Go package name, lowercase (e.g. fedtrust)")
    ap.add_argument("--tier", required=True, choices=list(TIERS), help="layering tier")
    ap.add_argument("--register", action="store_true", help="add to the defconfig (request-path driver)")
    ap.add_argument("--summary", default="", help="one-line package summary for the doc")
    ap.add_argument("--dry-run", action="store_true")
    args = ap.parse_args()

    name, tier = args.name, args.tier
    n = TIERS[tier]
    summary = args.summary or f"a tier-{tier} leaf (describe its single responsibility)"

    if not NAME_RE.match(name):
        print(f"ERROR: {name!r} is not a valid lowercase Go package name", file=sys.stderr)
        return 2
    if tier == "root":
        print("ERROR: 'root' is reserved for internal/abi; pick foundation or higher", file=sys.stderr)
        return 2

    root = repo_root()
    leaf_dir = root / "fak" / "internal" / name
    architest = root / "fak" / "internal" / "architest" / "architest_test.go"
    registrations = root / "fak" / "internal" / "registrations" / "registrations.go"
    dos_toml = root / "dos.toml"

    report: dict = {"name": name, "tier": tier, "register": args.register, "dry_run": args.dry_run, "edits": []}

    if leaf_dir.exists():
        print(f"ERROR: {leaf_dir} already exists — refusing to overwrite", file=sys.stderr)
        return 2

    arch_text = architest.read_text(encoding="utf-8")
    if f'"{name}":' in arch_text:
        print(f"ERROR: tier table already declares {name!r}", file=sys.stderr)
        return 2
    new_arch = insert_before_marker(arch_text, TIER_MARKER, f'\t"{name}": {n},\n')

    new_reg = None
    if args.register:
        reg_text = registrations.read_text(encoding="utf-8")
        new_reg = add_registration(reg_text, name)

    # Every leaf is a disjoint tree -> its own dos concurrency lane (keeps the
    # partition current so the fleet can edit it in parallel). Skipped if dos.toml
    # is absent in this checkout.
    new_dos = None
    if dos_toml.exists():
        new_dos = add_leaf_lane(dos_toml.read_text(encoding="utf-8"), name)

    # --- apply (or report) ---
    files = {
        leaf_dir / "doc.go": doc_go(name, tier, n, summary),
        leaf_dir / f"{name}.go": impl_go(name, args.register),
        leaf_dir / f"{name}_test.go": test_go(name),
    }
    if not args.dry_run:
        leaf_dir.mkdir(parents=True, exist_ok=False)
        for path, content in files.items():
            path.write_text(content, encoding="utf-8")
        architest.write_text(new_arch, encoding="utf-8")
        if new_reg is not None:
            registrations.write_text(new_reg, encoding="utf-8")
        if new_dos is not None:
            dos_toml.write_text(new_dos, encoding="utf-8")

    report["edits"] = [str(p.relative_to(root)) for p in files] + [
        str(architest.relative_to(root)) + " (tier table)"
    ] + ([str(registrations.relative_to(root)) + " (defconfig)"] if args.register else []) + (
        ["dos.toml (concurrency lane)"] if new_dos is not None else []
    )
    report["next_steps"] = [
        f"implement the leaf in fak/internal/{name}/{name}.go",
        f".\\fak\\test.ps1 ./internal/{name}/ ./internal/architest/   # both green by construction",
        "the architest gate now enforces this leaf's tier on every CI run",
    ]
    print(json.dumps(report, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: internal/newleaf/newleaf.go
package newleaf

import (
	"encoding/json"
	"fmt"
	"go/parser"
	"go/token"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

const (
	ModulePrefix = "github.com/anthony-chaudhary/fak/internal"
	TierMarker   = "// new-leaf:tier"
	LaneMarker   = "# new-leaf:lane"
	TreeMarker   = "# new-leaf:tree"
)

var (
	NameRE = regexp.MustCompile(`^[a-z][a-z0-9]*$`)
	Tiers  = map[string]int{"root": 0, "primitive": 1, "foundation-composite": 2, "mechanism": 3, "composer": 4, "integrator": 5}
)

type Options struct {
	Root     string
	Name     string
	Tier     string
	Register bool
	Summary  string
	DryRun   bool
}

type Report struct {
	Name         string   `json:"name"`
	Tier         string   `json:"tier"`
	Register     bool     `json:"register"`
	DryRun       bool     `json:"dry_run"`
	TierAdvisory string   `json:"tier_advisory,omitempty"`
	Edits        []string `json:"edits"`
	NextSteps    []string `json:"next_steps"`
}

func DocGo(name, tier string, n int, summary string) string {
	return fmt.Sprintf(
		"// Package %s is %s.\n//\n// Tier: %s (%d) - see internal/architest. This package may import only\n// packages whose tier is <= %d; an upward import fails the architest gate.\n// See AGENTS.md and internal/architest for the layering contract.\npackage %s\n",
		name, summary, tier, n, n, name,
	)
}

func ImplGo(name string, register bool) string {
	head := "package " + name + "\n"
	if register {
		head += "\nimport \"" + ModulePrefix + "/abi\"\n"
	}
	body := "\n// Ready reports that the leaf is wired. Replace this placeholder with the\n" +
		"// real surface this package exists to provide.\n" +
		"func Ready() bool { return true }\n"
	if register {
		body += "\n// init registers this leaf's driver against the frozen ABI before the\n" +
			"// kernel boots. Replace this placeholder with the real abi.Register* call.\n" +
			"func init() {\n\t_ = abi.ABIMinor\n}\n"
	}
	return head + body
}

func TestGo(name string) string {
	return "package " + name + "\n\n" +
		"import \"testing\"\n\n" +
		"// TestSpine drives the generated leaf's real surface end to end. Keep this\n" +
		"// representative path working while the proof envelope expands around it.\n" +
		"func TestSpine(t *testing.T) {\n" +
		"\tif !Ready() {\n" +
		"\t\tt.Fatal(\"generated leaf spine did not reach Ready\")\n" +
		"\t}\n" +
		"}\n"
}

// isMarkerLine reports whether ln IS the marker line rather than a line that
// merely MENTIONS the marker.
//
// The distinction is load-bearing, and a substring match got it wrong: dos.toml
// carries a prose comment documenting why some lanes landed by hand ("... so the
// `# new-leaf:lane` auto-insert never fired"), and that sentence sits INSIDE the
// [lanes].concurrent array. A Contains-based match treated it as a third marker,
// so every generated leaf was inserted into `concurrent` twice — once at the real
// marker, once above the prose — and the duplicate tripped the roster gate
// (CONCURRENT_LANES_OVERLAP, internal/architest TestLaneRosterHasNoDuplicates) on
// a tree the generator had just written. A real marker always OPENS its line, so
// anchoring the match to the trimmed prefix keeps documentation about the marker
// from acting as one.
func isMarkerLine(ln, marker string) bool {
	return strings.HasPrefix(strings.TrimSpace(ln), marker)
}

func InsertBeforeMarker(text, marker, line string) (string, error) {
	var out strings.Builder
	done := false
	for _, ln := range splitKeepLines(text) {
		if !done && isMarkerLine(ln, marker) {
			out.WriteString(line)
			done = true
		}
		out.WriteString(ln)
	}
	if !done {
		return "", fmt.Errorf("marker %q not found", marker)
	}
	return out.String(), nil
}

func InsertBeforeAllMarkers(text, marker, line string) (string, error) {
	var out strings.Builder
	hits := 0
	for _, ln := range splitKeepLines(text) {
		if isMarkerLine(ln, marker) {
			out.WriteString(line)
			hits++
		}
		out.WriteString(ln)
	}
	if hits == 0 {
		return "", fmt.Errorf("marker %q not found", marker)
	}
	return out.String(), nil
}

func AddRegistration(text, name string) (string, error) {
	imp := "\t_ \"" + ModulePrefix + "/" + name + "\"\n"
	if strings.Contains(text, imp) {
		return text, nil
	}
	idx := strings.LastIndex(text, "\n)")
	if idx < 0 {
		return "", fmt.Errorf("could not find import block close")
	}
	return text[:idx+1] + imp + text[idx+1:], nil
}

func AddLeafLane(text, name string) (string, error) {
	if strings.Contains(text, "[\"internal/"+name+"/**\"]") || strings.Contains(text, "fak/internal/"+name+"/**") {
		return text, nil
	}
	var err error
	text, err = InsertBeforeAllMarkers(text, LaneMarker, "  \""+name+"\",\n")
	if err != nil {
		return "", err
	}
	text, err = InsertBeforeMarker(text, TreeMarker, name+" = [\"internal/"+name+"/**\"]\n")
	if err != nil {
		return "", err
	}
	return text, nil
}

func Apply(opts Options) (Report, error) {
	name := opts.Name
	tier := opts.Tier
	n, ok := Tiers[tier]
	if !ok {
		return Report{}, fmt.Errorf("unknown tier %q", tier)
	}
	if !NameRE.MatchString(name) {
		return Report{}, fmt.Errorf("%q is not a valid lowercase Go package name", name)
	}
	if tier == "root" {
		return Report{}, fmt.Errorf("'root' is reserved for internal/abi; pick primitive or higher")
	}
	root := opts.Root
	if root == "" {
		var err error
		root, err = os.Getwd()
		if err != nil {
			return Report{}, err
		}
	}
	summary := opts.Summary
	if summary == "" {
		summary = "a tier-" + tier + " leaf (describe its single responsibility)"
	}

	leafDir := filepath.Join(root, "internal", name)
	if _, err := os.Stat(leafDir); err == nil {
		return Report{}, fmt.Errorf("%s already exists - refusing to overwrite", leafDir)
	}
	archPath := filepath.Join(root, "internal", "architest", "architest_test.go")
	regPath := filepath.Join(root, "internal", "registrations", "registrations.go")
	dosPath := filepath.Join(root, "dos.toml")

	archText, err := os.ReadFile(archPath)
	if err != nil {
		return Report{}, err
	}
	if strings.Contains(string(archText), "\""+name+"\":") {
		return Report{}, fmt.Errorf("tier table already declares %q", name)
	}
	newArch, err := InsertBeforeMarker(string(archText), TierMarker, fmt.Sprintf("\t%q: %d,\n", name, n))
	if err != nil {
		return Report{}, err
	}

	var newReg string
	if opts.Register {
		regText, err := os.ReadFile(regPath)
		if err != nil {
			return Report{}, err
		}
		newReg, err = AddRegistration(string(regText), name)
		if err != nil {
			return Report{}, err
		}
	}

	var newDos string
	if raw, err := os.ReadFile(dosPath); err == nil {
		newDos, err = AddLeafLane(string(raw), name)
		if err != nil {
			return Report{}, err
		}
	}

	files := map[string]string{
		filepath.Join(leafDir, "doc.go"):        DocGo(name, tier, n, summary),
		filepath.Join(leafDir, name+".go"):      ImplGo(name, opts.Register),
		filepath.Join(leafDir, name+"_test.go"): TestGo(name),
	}
	if !opts.DryRun {
		if err := os.MkdirAll(leafDir, 0o755); err != nil {
			return Report{}, err
		}
		for path, content := range files {
			if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
				return Report{}, err
			}
		}
		if err := os.WriteFile(archPath, []byte(newArch), 0o644); err != nil {
			return Report{}, err
		}
		if opts.Register {
			if err := os.WriteFile(regPath, []byte(newReg), 0o644); err != nil {
				return Report{}, err
			}
		}
		if newDos != "" {
			if err := os.WriteFile(dosPath, []byte(newDos), 0o644); err != nil {
				return Report{}, err
			}
		}
	}

	report := Report{Name: name, Tier: tier, Register: opts.Register, DryRun: opts.DryRun}
	for _, p := range []string{
		filepath.Join("internal", name, "doc.go"),
		filepath.Join("internal", name, name+".go"),
		filepath.Join("internal", name, name+"_test.go"),
	} {
		report.Edits = append(report.Edits, filepath.ToSlash(p))
	}
	report.Edits = append(report.Edits, "internal/architest/architest_test.go (tier table)")
	if opts.Register {
		report.Edits = append(report.Edits, "internal/registrations/registrations.go (defconfig)")
	}
	if newDos != "" {
		report.Edits = append(report.Edits, "dos.toml (concurrency lane)")
	}
	report.NextSteps = []string{
		"1. name the end-to-end outcome and implement its smallest applied path through the real seam in internal/" + name + "/" + name + ".go",
		"2. go test ./internal/" + name + " ./internal/architest to capture the working spine test or command that drives the real " + name + " object end to end",
		"3. fak-dev issue fanout --title \"<feature>\" --leaf " + name + " --spine <commit|command|doc> --json",
		"4. expand the exhaustive proof envelope around that spine: failure paths, edge cases, platforms, concurrency, and soak",
		"5. measure the end-to-end baseline before optimizing, then keep only a net-true gain",
	}
	// Nudge toward the minimum-correct tier at creation (#4045): the skeleton
	// imports only abi (when registered) or nothing, so declaring above foundation
	// surfaces an advisory the moment the leaf is scaffolded — foundation stops
	// being the frictionless default without any creation being blocked.
	var scaffoldDeps []string
	if opts.Register {
		scaffoldDeps = []string{"abi"}
	}
	report.TierAdvisory = TierAdvisory(tier, scaffoldDeps, ParseTierTable(string(archText)))
	return report, nil
}

func (r Report) JSON() ([]byte, error) {
	return json.MarshalIndent(r, "", "  ")
}

// --- minimum-correct-tier suggestion (#4045) --------------------------------
//
// The imbalance the architest-foundation epic (#4041) chases is authored in one
// `new-leaf` call at a time: with no signal about what a leaf imports, the path of
// least resistance is always `--tier foundation`. These helpers compute the minimum
// tier consistent with the layered-DAG rule (a package may import only packages of
// tier <= its own, so its floor is max(tier(dep))) directly from the leaf's imports,
// so the author chooses from evidence. Advisory only — edit-time enforcement is #2082.

var tierRowRE = regexp.MustCompile(`"([a-z][a-z0-9]*)"\s*:\s*(\d+)`)

// TierNameForLevel reverses Tiers: a level (0..5) to its canonical tier name, or ""
// for an unknown level.
func TierNameForLevel(level int) string {
	for name, n := range Tiers {
		if n == level {
			return name
		}
	}
	return ""
}

// ParseTierTable extracts the package -> tier-level map from the architest tier
// table text (the `"name": N,` rows new-leaf itself maintains behind TierMarker).
// It is the read half of the table Apply writes into, so a suggestion is always
// computed against the same source of truth the gate enforces.
func ParseTierTable(archText string) map[string]int {
	out := map[string]int{}
	for _, m := range tierRowRE.FindAllStringSubmatch(archText, -1) {
		if n, err := strconv.Atoi(m[2]); err == nil {
			out[m[1]] = n
		}
	}
	return out
}

// ScanInternalDeps returns the sorted, unique internal leaf names imported by the
// non-test Go files in dir: an import of ".../internal/foo" or ".../internal/foo/bar"
// both yield "foo". It is the import scan the tier suggestion reasons over.
func ScanInternalDeps(dir string) ([]string, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}
	seen := map[string]struct{}{}
	fset := token.NewFileSet()
	for _, e := range entries {
		nm := e.Name()
		if e.IsDir() || !strings.HasSuffix(nm, ".go") || strings.HasSuffix(nm, "_test.go") {
			continue
		}
		f, err := parser.ParseFile(fset, filepath.Join(dir, nm), nil, parser.ImportsOnly)
		if err != nil {
			return nil, err
		}
		for _, imp := range f.Imports {
			p, err := strconv.Unquote(imp.Path.Value)
			if err != nil {
				continue
			}
			rest, ok := strings.CutPrefix(p, ModulePrefix+"/")
			if !ok {
				continue
			}
			leaf := rest
			if i := strings.IndexByte(leaf, '/'); i >= 0 {
				leaf = leaf[:i]
			}
			if leaf != "" {
				seen[leaf] = struct{}{}
			}
		}
	}
	out := make([]string, 0, len(seen))
	for k := range seen {
		out = append(out, k)
	}
	sort.Strings(out)
	return out, nil
}

// MinTier returns the minimum legal tier for a leaf with the given internal dependencies.
// No dependencies (or abi alone) is primitive. Any other internal dependency makes the leaf
// at least foundation-composite; a still-higher dependency raises that floor.
func MinTier(deps []string, tierOf map[string]int) (level int, governing string) {
	level = Tiers["primitive"]
	for _, d := range deps {
		if d != "abi" && level < Tiers["foundation-composite"] {
			level, governing = Tiers["foundation-composite"], d
		}
		if t, ok := tierOf[d]; ok && t > level {
			level, governing = t, d
		}
	}
	return level, governing
}

// TierAdvisory compares the author-declared tier against the minimum legal tier
// derived from the deps and returns a one-line advisory when they differ, or "" when
// the declared tier is already minimum-correct. Advisory only — never a block.
func TierAdvisory(declared string, deps []string, tierOf map[string]int) string {
	dl, ok := Tiers[declared]
	if !ok {
		return ""
	}
	level, gov := MinTier(deps, tierOf)
	name := TierNameForLevel(level)
	switch {
	case dl > level:
		if gov == "" {
			return fmt.Sprintf("tier advisory: imports only tier-≤%d packages → %q suffices (you declared %q)", level, name, declared)
		}
		return fmt.Sprintf("tier advisory: highest-tier import %q is tier-%d → %q suffices (you declared %q)", gov, level, name, declared)
	case dl < level:
		return fmt.Sprintf("tier advisory: imports %q (tier-%d) → tier must be ≥ %q (you declared %q); the architest gate rejects an upward import", gov, level, name, declared)
	}
	return ""
}

// Suggestion is the read-only minimum-correct-tier verdict for an existing leaf.
type Suggestion struct {
	Leaf           string   `json:"leaf"`
	DeclaredTier   string   `json:"declared_tier,omitempty"`
	SuggestedTier  string   `json:"suggested_tier"`
	SuggestedLevel int      `json:"suggested_level"`
	GoverningDep   string   `json:"governing_dep,omitempty"`
	InternalDeps   []string `json:"internal_deps"`
	Advisory       string   `json:"advisory,omitempty"`
}

// Suggest computes the minimum-correct tier for an existing internal leaf from its
// imports, read-only. A non-empty declared tier adds a comparison advisory. root
// defaults to the current working directory.
func Suggest(root, leaf, declared string) (Suggestion, error) {
	if !NameRE.MatchString(leaf) {
		return Suggestion{}, fmt.Errorf("%q is not a valid lowercase Go package name", leaf)
	}
	if declared != "" {
		if _, ok := Tiers[declared]; !ok {
			return Suggestion{}, fmt.Errorf("unknown tier %q", declared)
		}
	}
	if root == "" {
		var err error
		root, err = os.Getwd()
		if err != nil {
			return Suggestion{}, err
		}
	}
	archText, err := os.ReadFile(filepath.Join(root, "internal", "architest", "architest_test.go"))
	if err != nil {
		return Suggestion{}, err
	}
	tierOf := ParseTierTable(string(archText))
	deps, err := ScanInternalDeps(filepath.Join(root, "internal", leaf))
	if err != nil {
		return Suggestion{}, err
	}
	level, gov := MinTier(deps, tierOf)
	if deps == nil {
		deps = []string{}
	}
	s := Suggestion{
		Leaf:           leaf,
		DeclaredTier:   declared,
		SuggestedTier:  TierNameForLevel(level),
		SuggestedLevel: level,
		GoverningDep:   gov,
		InternalDeps:   deps,
	}
	if declared != "" {
		s.Advisory = TierAdvisory(declared, deps, tierOf)
	}
	return s, nil
}

// JSON renders the suggestion as indented JSON.
func (s Suggestion) JSON() ([]byte, error) {
	return json.MarshalIndent(s, "", "  ")
}

func splitKeepLines(s string) []string {
	if s == "" {
		return nil
	}
	parts := strings.SplitAfter(s, "\n")
	if parts[len(parts)-1] == "" {
		parts = parts[:len(parts)-1]
	}
	return parts
}
